File: java/src/com/alofmethbin/jadawin/elements/Table.java
package com.alofmethbin.jadawin.elements;

import com.alofmethbin.jadawin.Article;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Table extends Element {
    private List<String> columns = new ArrayList<>();
    private List<List<Text>> rows = new ArrayList<>();
    
    public Table( Article article, List<String> lines) throws IOException {
        super( article);
        
        for (String column: divide( lines.get(0))) {
            columns.add( checkLabel( column));
        }
        
        for (int i = 1; i < lines.size(); i++) {
            List<Text> row = new ArrayList<>();
            rows.add( row);
            
            for (String value: divide( lines.get(i))) {
                row.add( new Text( article, value));
            }
        }
    }
    
    private String [] divide( String line) {
        line = line.trim();
        if ( line.startsWith( "|") ) {line = line.substring(1);}
        if ( line.endsWith( "|") ) {line = line.substring( 0, line.length() - 1);}
        return line.split( "\\|");
    }

    @Override
    public int lineCount() {
        return 5 + rows.size();
    }

    @Override
    public void prepare() {
        for (List<Text> row: rows) {
            for (Text text: row) {text.prepare();} 
        }
    }
}

File: ruby/elements/table.rb
require_relative 'Text'

module Elements
  class Table < Element
    include Utils

    def initialize( compiler, article, lines)
      @colums = @rows = []

      lines = lines.collect do |line|
        if m = /^\|(.*)$/.match( line)
          line = m[1]
        end
        if m = /^(.*)\|\s*$/.match( line)
          line = m[1]
        end
        line
      end

      @columns = lines[0].split( '|').collect do |column|
        check_label( article, column)
      end

      @rows = lines[1..-1].collect do |line|
        line.split('|').collect do |field|
          Text.new( compiler, article, (field.strip != '') ? [field] : [])
        end
      end
    end

    def columns
      @columns.collect do |column|
        {'data' => [{'type' => 'raw', 'text' => column}], 'bold' => true, 'align' => 'center'}
      end
    end

    def is_number?( markup)
      return false unless markup.size == 1
      /^\d+$/ =~ markup[0]['text']
    end

    def line_count
      5 + @rows.size
    end

    def prepare( compiler, article, parents, younger)
      @rows.each do |row|
        row.each do |column|
          column.prepare( compiler, article, parents, younger)
        end
      end
    end

    def row( compiler, article, r)
      fields = r.collect do |field|
        markup = field.to_data_single_paragraph( compiler, article)
        align = is_number?( markup) ? 'right' : 'left'
        {'data' => markup, 'bold' => false, 'align' => align}
      end

      while fields.size < @columns.size
        fields << {'data' => [{'type' => 'none'}], 'bold' => false, 'align' => 'left'}
      end

      fields
    end

    def rows( compiler, article)
      @rows.collect {|r| row( compiler, article, r)}
    end

    def to_data( compiler, article)
      {'type'  => 'table',
       'rows'  => [columns] + rows( compiler, article)}
    end
  end
end